controllers/imagepolicy_controller.go

```diff
@@ -39,17 +39,14 @@ import (
 	"github.com/fluxcd/pkg/version"
 
 	imagev1alpha1 "github.com/fluxcd/image-reflector-controller/api/v1alpha1"
+	"github.com/fluxcd/image-reflector-controller/internal/database"
 )
 
 // this is used as the key for the index of policy->repository; the
 // string is arbitrary and acts as a reminder where the value comes
 // from.
 const imageRepoKey = ".spec.imageRepository.name"
 
-type DatabaseReader interface {
```

Comment:
Interfaces generally belong at the location of the requirement, rather than that of the (an) implementation. If it feels off to use a definition in `imagepolicy_controller.go` from `imagerepository_controller.go`, you could put both the definitions in database.go (back in this directory).

Reply:
I'm not sure I totally agree with this, while I can understand your thinking, and it comes directly from the Go Code Review guide https://github.com/golang/go/wiki/CodeReviewComments#interfaces and I'd argue that it's useful

It also feels like BadgerDatabase is defining methods on a value that aren't meeting any requirement, and the only place that requirement is defined is in the controller, which I personally find odd.

https://rakyll.org/style-packages/ says...

> In some cases, you may not be able to provide all related types from a single package. It might be noisy to do so, or you might want to publish concrete implementations of a common interface from a separate package, or those types could be owned by a third-party package. 

Which implies that it's common to publish implementations with a common interface in the single package.

Reply:
> It also feels like BadgerDatabase is defining methods on a value that aren't meeting any requirement, and the only place that requirement is defined is in the controller

Yes, it feels like a trapeze act, I know what you mean. It's a shame the CodeReviewComments wiki page doesn't really explain the _why_ of its advice. The way I think about it is this: you use an interface to decouple your use of a value from wherever the value comes from. If you put the interface with the implementation, you reintroduce that coupling.

Thanks for harmonising with my preferences here 🤓